docs/proposals/20220221-runtime-SDK.md

```diff
@@ -0,0 +1,802 @@
+
+In other words, Runtime Hooks are not concerned about “low-level” details of how Kubernetes handles objects in the
+API Server/etcd; Runtime Hooks instead focus on “high-level” events of a Cluster’s lifecycle, or _by opportunity_,
+in the Cluster API reconcile loops.
+
+Please note that, no matter the similarities in some part of the design, users should not make assumptions about
+Runtime Hooks having properties or behaviors typical of Kubernetes admission webhooks unless they are explicitly
+defined in the following paragraphs.
+
+#### Runtime SDK rules
+
+As this proposal is based on RESTful APIs, we are using [OpenAPI Specification v3.0.0](https://swagger.io/specification/) [1] 
+to document Runtime Hooks supported by Cluster API.
+
+Most specifically, a single OpenAPI document providing specification for all the Runtime Hooks supported by a
+Cluster API release will be added to the release artifacts; users can rely on https://editor.swagger.io/ or similar
+tools to view the specification, and during implementation we will consider adding similar view to the Cluster API
+book as well e.g.
+
+![overview](images/runtime-sdk/swagger-ui.png)
+
+Each Runtime Hook will be defined by one (or more) RESTful APIs implemented as a `POST` operation; each operation
+is going to receive an input parameter as a request body, and return an output value as response body, both
+`application/json` encoded and with a schema of arbitrary complexity that should be considered an integral part of
+the Runtime Hook definition.
+
+It is also worth noting that more than one version of the same Runtime Hook might be supported at the same time;
+e.g. in the example above the `Cluster.BeforeUpgrade` operations exist in version `v1alpha1` (old version)
+and `v1alpha2` (current).
+
+Supporting more versions at the same time is a requirement in order to:
+
+- Allow Cluster API maintainers to continue to develop and evolve Runtime Hooks in a predictable way.
+- Provide a well-defined set of guarantees to Runtime Extension implementers they can rely on while developing
+  solutions on top of CAPI.
+
+In their simplest form guarantees for Runtime Hooks versions are:
+
+- Once a Runtime Hook version has been published, breaking changes are not allowed without bumping to a new version
+  (e.g. fields removal/renaming)
+- Before removing a Runtime Hook version, a deprecation period should be respected, with a duration depending
+  on the maturity of the API itself (12 Months/3 Versions GA, 6M/2V beta, 0 alpha).
+
+The formal definition of the Runtime SDK rules derived from https://kubernetes.io/docs/reference/using-api/deprecation-policy/,
+can be found in the annex at the end of the document. Please note that during implementation we will consider a
+mechanism allowing to:
+
+- Inform admins about Runtime Extension using a deprecated version of a Runtime Hook (e.g. return a well known
```

Comment:
Added as an option.
However, webhook warnings apply does not alert you about existing extensions being deprecated unless you change the API object

Reply:
That's correct. For that it would be good to have warnings in clusterctl or maybe even on Provider CR create/update.

I'm just not sure a HTTP header somewhere would be noticed. But I'm just not really aware if there are well-known headers for something like that.